redwood-layout-shared-test/src/commonMain/kotlin/app/cash/redwood/layout/AbstractBoxTest.kt

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (C) 2024 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package app.cash.redwood.layout
+
+import app.cash.redwood.layout.api.Constraint
+import app.cash.redwood.layout.widget.Box
+import app.cash.redwood.ui.Dp
+import app.cash.redwood.ui.dp
+import kotlin.test.Test
+
+abstract class AbstractBoxTest<T : Any> {
+
+  abstract fun Box(block: Box<T>.() -> Unit = {}): Box<T>
+
+  abstract fun Color(block: Color<T>.() -> Unit = {}): Color<T>
+
+  fun Color(color: Int, width: Dp, height: Dp) = Color {
+    color(color)
+    width(width)
+    height(height)
+  }
+
+  abstract fun verifySnapshot(value: T)
+
+  @Test
+  open fun testDefaults() {
+    val widget = Box()
+    verifySnapshot(widget.value)
+  }
+
+  @Test
+  open fun testWrap() {
+    val widget = Box {
+      width(Constraint.Wrap)
+      height(Constraint.Wrap)
+    }
+    verifySnapshot(widget.value)
+  }
+
+  @Test
+  open fun testFill() {
+    val widget = Box {
+      width(Constraint.Fill)
+      height(Constraint.Fill)
+    }
+    verifySnapshot(widget.value)
+  }
+
+  @Test
+  open fun testWrapWithChildren() {
```

Comment:
For this and the test below, can you make each test body a helper function and then make four test cases for the four alignments (top-left, top-right, bottom-left, bottom-right)? I can also do as a follow-up.

Reply:
There's also `CENTER` so we'd end up with 2 x 9 tests. I can add them manually, but is there a better way to parameterize these tests?

Reply:
Nope!

We gotta build a parameterized Kotlin compiler plugin so they work on non-JVM.